File: apps/ios/Packages/TrackEvolutionKit/Tests/TrackEvolutionKitTests/RepoRoot.swift
import Foundation

/// Locates the repository root by walking up from this source file.
///
/// The contract fixtures (`contracts/golden/`, `contracts/logic/`) are read from
/// the repo rather than copied into the package as bundled resources: a copy
/// would go stale exactly when it mattered, which is the opposite of what those
/// fixtures are for.
enum RepoRoot {
    static let url: URL = {
        var dir = URL(fileURLWithPath: #filePath).deletingLastPathComponent()
        while dir.path != "/" {
            if FileManager.default.fileExists(atPath: dir.appending(path: "package.json").path) {
                return dir
            }
            dir = dir.deletingLastPathComponent()
        }
        fatalError("repository root not found above \(#filePath)")
    }()

    static func path(_ relative: String) -> URL {
        url.appending(path: relative)
    }
}

File: apps/android/core/src/test/kotlin/app/trackevolution/core/RepoRoot.kt
package app.trackevolution.core

import java.io.File

/**
 * Locates the repository root by walking up from the Gradle working directory.
 *
 * The contract fixtures (`contracts/golden/`) are read from the repo rather than
 * copied into the module's test resources: a copy would go stale exactly when it
 * mattered, which is the opposite of what those fixtures are for. The iOS kit
 * does the same thing in its own `RepoRoot`.
 */
object RepoRoot {
    val dir: File = run {
        var candidate: File? = File(System.getProperty("user.dir")).absoluteFile
        while (candidate != null) {
            if (File(candidate, "package.json").isFile) return@run candidate
            candidate = candidate.parentFile
        }
        error("repository root not found above ${System.getProperty("user.dir")}")
    }

    fun path(relative: String): File = File(dir, relative)
}
